In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt
import sklearn
from sklearn.model_selection import train_test_split  
from IPython.display import display   ## To display function results without needing a Return

from sklearn.metrics import mutual_info_score

from sklearn.feature_extraction import DictVectorizer

from sklearn.linear_model import LogisticRegression

In [4]:
!pip install wget

  Preparing metadata (setup.py) ... done
  DEPRECATION: Building 'wget' using the legacy setup.py bdist_wheel mechanism, which will be removed in a future version. pip 25.3 will enforce this behaviour change. A possible replacement is to use the standardized build interface by setting the `--use-pep517` option, (possibly combined with `--no-build-isolation`), or adding a `pyproject.toml` file to the source tree of 'wget'. Discussion can be found at https://github.com/pypa/pip/issues/6334
  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9685 sha256=d047707a35132f46e5448b57770be2ff2155d168aa8b9345d86f152759a16656
  Stored in directory: /home/codespace/.cache/pip/wheels/40/b3/0f/a40dbd1c6861731779f62cc4babcb234387e11d697df70ee97
Successfully built wget


In [2]:
## USE WGET TO DOWNLOAD FILE

# 2. Define the URL of the CSV file you want to download
file_url = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/course_lead_scoring.csv'

# 3. Define the local filename to save the downloaded file
filename = 'course_lead_scoring.csv'

# 4. Use `wget` with an exclamation mark to download the file
!wget -O {filename} {file_url}

--2025-10-13 20:06:52--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/course_lead_scoring.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 80876 (79K) [text/plain]
Saving to: ‘course_lead_scoring.csv’

course_lead_scoring 100%[===================>]  78.98K  --.-KB/s    in 0.01s   

2025-10-13 20:06:52 (7.50 MB/s) - ‘course_lead_scoring.csv’ saved [80876/80876]



In [3]:
df = pd.read_csv(filename)

df.head(2)

lead_source industry  number_of_courses_viewed  annual_income  \
0      paid_ads      NaN                         1        79450.0   
1  social_media   retail                         1        46992.0   

  employment_status       location  interaction_count  lead_score  converted  
0        unemployed  south_america                  4        0.94          1  
1          employed  south_america                  1        0.80          0

In [4]:
## QUESTION 1 What is the most frequent observation (mode) for the column industry?

df.industry.mode()

0    retail
Name: industry, dtype: object

### QUESTION 2 

Create the correlation matrix for the numerical features of your dataset. In a correlation matrix, you compute the correlation coefficient between every pair of features.

What are the two features that have the biggest correlation?

In [11]:
df.dtypes

lead_source                  object
industry                     object
number_of_courses_viewed      int64
annual_income               float64
employment_status            object
location                     object
interaction_count             int64
lead_score                  float64
converted                     int64
dtype: object

In [21]:
## SAVE NUMERICAL COLUMNS INTO LIST
    ## FILTER DATAFRAME BY SAID LIST AFTERWARDS
numerical = list(df.dtypes[df.dtypes.isin(['int64', 'float64'])].index)


In [22]:
numerical

['number_of_courses_viewed',
 'annual_income',
 'interaction_count',
 'lead_score',
 'converted']

In [32]:
type(df_number)

method

In [35]:
df_number = df[numerical]

df_number

number_of_courses_viewed  annual_income  interaction_count  lead_score  \
0                            1        79450.0                  4        0.94   
1                            1        46992.0                  1        0.80   
2                            5        78796.0                  3        0.69   
3                            2        83843.0                  1        0.87   
4                            3        85012.0                  3        0.62   
...                        ...            ...                ...         ...   
1457                         1            NaN                  4        0.53   
1458                         3        65259.0                  2        0.24   
1459                         1        45688.0                  3        0.02   
1460                         5        71016.0                  0        0.25   
1461                         3        92855.0                  3        0.41   

      converted  
0             1  
1             0  
2             1  
3             0  
4             1  
...         ...  
1457          1  
1458          1  
1459          1  
1460          1  
1461          1  

[1462 rows x 5 columns]

In [43]:
df_number.corr().abs()

number_of_courses_viewed  annual_income  \
number_of_courses_viewed                  1.000000       0.031551   
annual_income                             0.031551       1.000000   
interaction_count                         0.023565       0.048618   
lead_score                                0.004879       0.005334   
converted                                 0.435914       0.078256   

                          interaction_count  lead_score  converted  
number_of_courses_viewed           0.023565    0.004879   0.435914  
annual_income                      0.048618    0.005334   0.078256  
interaction_count                  1.000000    0.009888   0.374573  
lead_score                         0.009888    1.000000   0.193673  
converted                          0.374573    0.193673   1.000000

What are the two features that have the biggest correlation?

interaction_count and lead_score  --> 0.0009888

number_of_courses_viewed and lead_score --> 0.004879

number_of_courses_viewed and interaction_count  --> 0.023565

annual_income and interaction_count  --> 0.048618

Only consider the pairs above when answering this question.  

ANSWER = The 2 highest correlations are shared between {"number_of_courses_viewed" and "interaction_count"} 

and {"number_of_courses_viewed" and "interaction_count"}

